File: integrations/pqc-experiment/microbench.py
import sys,json,time
import cryptography
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives import serialization
from vella import proof_v2 as base
import proof as pq
r=json.load(sys.stdin);record=r['record'];rows=[]
def q(values):
    s=sorted(values);return {'n':len(s),'p50':s[int(len(s)*.5)],'p95':s[int(len(s)*.95)],'p99':s[int(len(s)*.99)],'max':s[-1]}
for suite in ['baseline',*pq.PROFILES]:
    if suite=='baseline':
        k=ec.generate_private_key(ec.SECP256R1());private=k.private_bytes(serialization.Encoding.PEM,serialization.PrivateFormat.PKCS8,serialization.NoEncryption());public=k.public_key().public_bytes(serialization.Encoding.PEM,serialization.PublicFormat.SubjectPublicKeyInfo)
        sign=lambda:base.sign_v2(record,private);verify=lambda b:base.verify_v2(b,public)
    else:
        keys=pq.generate_keys(suite);sign=lambda:pq.sign(record,keys['privateKeys'],suite);verify=lambda b:pq.verify(b,keys['publicKeys'],suite)
    for _ in range(100):assert verify(sign())['ok']
    signing=[];verification=[]
    for _ in range(2000):
        t=time.perf_counter_ns();bundle=sign();signing.append((time.perf_counter_ns()-t)/1e6)
        t=time.perf_counter_ns();v=verify(bundle);verification.append((time.perf_counter_ns()-t)/1e6);assert v['ok']
    rows.append({'runtime':'Python','python':sys.version.split()[0],'cryptography':cryptography.__version__,'suite':suite,'contentBytes':r['size'],'signMs':q(signing),'verifyMs':q(verification),'envelopeBytes':len(json.dumps(bundle,separators=(',',':')).encode())})
json.dump(rows,sys.stdout)

File: integrations/pqc-experiment/microbench.mjs
import {performance} from 'node:perf_hooks';
import {generateKeyPairSync} from 'node:crypto';
import {spawnSync} from 'node:child_process';
import {mkdirSync,writeFileSync} from 'node:fs';
import {resolve,dirname} from 'node:path';
import {fileURLToPath} from 'node:url';
import {cpus} from 'node:os';
import assert from 'node:assert/strict';
import pq from './proof.cjs';
import base from '../../sdk/node/proof-v2.cjs';
import {createGovernor} from '../../sdk/node/governor.js';
const here=dirname(fileURLToPath(import.meta.url)),root=resolve(here,'../..'),output=resolve(process.argv[2]);
const q=values=>{const s=[...values].sort((a,b)=>a-b);return {n:s.length,p50:s[Math.floor(s.length*.5)],p95:s[Math.floor(s.length*.95)],p99:s[Math.floor(s.length*.99)],max:s.at(-1)};};
const rows=[];
for(const size of [1024,4096]){
 const ec=generateKeyPairSync('ec',{namedCurve:'prime256v1',privateKeyEncoding:{type:'pkcs8',format:'pem'},publicKeyEncoding:{type:'spki',format:'pem'}});
 const b=createGovernor().govern({intent:'EXECUTE_CHANGE',evidenceMask:1,action:{args:{text:'x'.repeat(size)}},proof:{signingKey:ec.privateKey}}).proofBundle;
 const record=base.verifyV2(b,ec.publicKey).authenticated;
 for(const suite of ['baseline',...Object.keys(pq.PROFILES)]){
  const keys=suite==='baseline'?ec:pq.generateKeys(suite);const sign=()=>suite==='baseline'?base.signV2(record,keys.privateKey):pq.sign(record,keys.privateKeys,suite);const verify=b=>suite==='baseline'?base.verifyV2(b,keys.publicKey):pq.verify(b,keys.publicKeys,suite);
  for(let i=0;i<100;i++)assert.equal(verify(sign()).ok,true);
  const signing=[],verification=[];let bundle;
  for(let i=0;i<2000;i++){let t=performance.now();bundle=sign();signing.push(performance.now()-t);t=performance.now();const v=verify(bundle);verification.push(performance.now()-t);assert.equal(v.ok,true);}
  rows.push({suite,contentBytes:size,signMs:q(signing),verifyMs:q(verification),envelopeBytes:Buffer.byteLength(JSON.stringify(bundle))});
 }
 const py=spawnSync(process.env.VELLA_VERIFY_PYTHON||'python3',[resolve(here,'microbench.py')],{input:JSON.stringify({record,size}),encoding:'utf8',env:{...process.env,PYTHONPATH:resolve(root,'sdk/python')}});assert.equal(py.status,0,py.stderr);rows.push(...JSON.parse(py.stdout));
}
mkdirSync(dirname(output),{recursive:true});writeFileSync(output,JSON.stringify({kind:'vella-pqc-microbench-v1',time:new Date().toISOString(),node:process.version,openssl:process.versions.openssl,cpu:cpus()[0].model,scope:'Sequential warm microbenchmarks; key generation excluded; P-256 PEM parsing and ML-DSA seed expansion included; not MCP workload acceptance',rows},null,2)+'\n');console.log(JSON.stringify(rows));
